src/server/game/Entities/Player/Player.cpp

```diff
@@ -11066,61 +11073,87 @@ InventoryResult Player::CanStoreItems(Item** items, int count, uint32* itemLimit
             for (uint8 t = KEYRING_SLOT_START; t < KEYRING_SLOT_END; ++t)
             {
                 item2 = GetItemByPos(INVENTORY_SLOT_BAG_0, t);
-                if (item2 && item2->CanBeMergedPartlyWith(pProto) == EQUIP_ERR_OK && keyringCounts[t-KEYRING_SLOT_START] + item->GetCount() <= pProto->GetMaxStackSize())
+                if (item2 && item2->CanBeMergedPartlyWith(pProto) == EQUIP_ERR_OK && keyringCounts[t-KEYRING_SLOT_START] < pProto->GetMaxStackSize())
                 {
-                    keyringCounts[t-KEYRING_SLOT_START] += item->GetCount();
-                    b_found = true;
-                    break;
+                    keyringCounts[t-KEYRING_SLOT_START] += remaining_count;
+                    remaining_count = keyringCounts[t-KEYRING_SLOT_START] - pProto->GetMaxStackSize();
+
+                    b_found = remaining_count <= 0;
+
+                    // stop checking keyring?
+                    if (b_found)
+                        break;
                 }
             }
+
             if (b_found)
                 continue;
 
             for (int t = CURRENCYTOKEN_SLOT_START; t < CURRENCYTOKEN_SLOT_END; ++t)
             {
                 item2 = GetItemByPos(INVENTORY_SLOT_BAG_0, t);
-                if (item2 && item2->CanBeMergedPartlyWith(pProto) == EQUIP_ERR_OK && currencyCounts[t-CURRENCYTOKEN_SLOT_START] + item->GetCount() <= pProto->GetMaxStackSize())
+                if (item2 && item2->CanBeMergedPartlyWith(pProto) == EQUIP_ERR_OK && currencyCounts[t-CURRENCYTOKEN_SLOT_START] < pProto->GetMaxStackSize())
                 {
-                    currencyCounts[t-CURRENCYTOKEN_SLOT_START] += item->GetCount();
-                    b_found = true;
-                    break;
+                    currencyCounts[t-CURRENCYTOKEN_SLOT_START] += remaining_count;
+                    remaining_count = currencyCounts[t-CURRENCYTOKEN_SLOT_START] - pProto->GetMaxStackSize();
```

Comment:
This and all similar checks are probably a source of problems when it doesn't form a complete stack. It is probably the source of my problems with Test 2.

e.g.

1. 5 items added to a 10-stack of a 20-stackable item, forming a 15-stack.
2. `remaining_count` is calculated as 15-20. Since this is a `uint32`, this overflows.

Reply:
While this was **a** problem that needed to be solved, it wasn't **the** problem.

I now think that the problem is that we are continually polling `GetItemByPos` to get the state of the item.

1. The first stack's iteration goes to the free-space check and says "This is a 0-stack slot, it's available to throw our items into.
2. The second stack's iteration:
    a. The merge check relies on `GetItemByPos` succeeding. Since the mock-inventory has a new stack that doesn't match the real inventory, this fails and the slot isn't considered for merging.
    b. The free-space check fails because it detects the 1-stack.

Solution:

1. When compiling the mock-inventory, we also need an array of `Item` pointers. This array shall be used in merge checks.
2. When using free-space, assign the value of `item` to the array of item pointers.

Reply:
Replacing `GetItemByPos` with another array of Item pointers solved the issue.

Successfully performed all tests using the latest build.